src/Pact/Typechecker.hs

```diff
@@ -541,9 +542,16 @@ unifyTypes l r = case (l,r) of
           Just (Left uc) -> Just $ vWrap $ TyVar $ v { _tvConstraint = uc }
           Just (Right uc) -> Just $ sWrap $ TyVar $ sv { _tvConstraint = uc }
           Nothing -> Nothing
-        (TypeVar _ ac,_) | null ac || s `elem` ac -> Just useS
+        (TypeVar _ ac,_) | checkConstraints s ac -> Just useS
         _ -> Nothing
 
+-- | check for unification of (non-var) type against constraints
+checkConstraints :: Eq n => Type n -> [Type n] -> Bool
+checkConstraints _ [] = True
+checkConstraints ty constrs = foldl' check False constrs where
```

Comment:
If performance mattered, I think `foldr` is the better choice here because it affords short-circuiting from the fold. But this is fine.

Reply:
arrgh missed this b4 merging, i'll fold it in later. 

(I'm here all week)